examples/larger-examples/datalog/ford-fulkerson.flix

```diff
@@ -0,0 +1,649 @@
+    ///
+    def minCapacity(p: Path[a], g: m[FlowNetwork[a]]): Int32 with Foldable[m], Order[a] =
+        let onPath = (s, d) -> isForwardEdge(src = s, dst = d, p) or isBackEdge(src = s, dst = d, p);
+        let optMin = g |> Foldable.filter(match (s, d, _, _) -> onPath(s, d))
+            |> List.map(match (_, _, u, f) -> u - f)
+            |> List.minimum;
+        match optMin {
+            case Some(u) => u
+            case None    => unreachable!() // This function is only called by `maxFlow` if an augmenting path was found
+        }
+
+    ///
+    /// Returns a new flow network where the edges in `g` on the `Path` `p` has been adjusted by `incr`.
+    ///
+    def increaseFlow(p: Path[a], incr: Flow, g: m[FlowNetwork[a]]): Vector[FlowNetwork[a]] with Foldable[m], Order[a] =
+        g
+        |> Foldable.toVector
+        |> Vector.map(match (s, d, u, f) ->
+            if (isForwardEdge(src = s, dst = d, p))
+                (s, d, u, f + incr)
+            else if (isBackEdge(src = s, dst = d, p))
+                (s, d, u, f - incr)
+            else
+                (s, d, u, f)
+        )
+
+    ///
+    /// Returns true if `src` is an edge pointing to `dst` on the `Path` `p`.
+    ///
+    def isForwardEdge(src: { src = a }, dst: { dst = a }, p: Path[a]): Bool with Eq[a] =
+        match (indexOf(src.src, p), indexOf(dst.dst, p)) { // A path is sorted in reverse order
+            case (Some(si), Some(di)) if di + 1 == si => true
+            case _ => false
+        }
+
+    ///
+    /// Returns true if `dst` is an edge pointing to `src` on the `Path` `p`.
+    ///
+    def isBackEdge(src: { src = a }, dst: { dst = a }, p: Path[a]): Bool with Eq[a] =
+        match (indexOf(src.src, p), indexOf(dst.dst, p)) { // A path is sorted in reverse order
+            case (Some(si), Some(di)) if si + 1 == di => true
+            case _ => false
+        }
+
+    pub enum Path[a] with ToString {
+        case Path(List[a])
+        case Bot // Infinitely long path
+    }
+
+    instance Eq[Path[a]] {
+        pub def eq(x: Path[a], y: Path[a]): Bool = match (x, y) {
+            case (Path(xs), Path(ys)) => List.length(xs) == List.length(ys)
+            case (Bot, _)             => false
+            case (_, Bot)             => false
+        }
+    }
+
+    instance Order[Path[a]] {
+        pub def compare(x: Path[a], y: Path[a]): Comparison = match (x, y) {
+            case (Path(xs), Path(ys)) => List.length(xs) <=> List.length(ys)
+            case (Bot, _)             => Comparison.GreaterThan
+            case (_, Bot)             => Comparison.GreaterThan
+        }
+    }
+
+    instance LowerBound[Path[a]] {
+        // The longest list
+        pub def minValue(): Path[a] = Bot
+    }
+
+    instance PartialOrder[Path[a]] {
+        pub def lessEqual(x: Path[a], y: Path[a]): Bool = match (x, y) {
+            case (Path(xs), Path(ys)) => List.length(xs) >= List.length(ys)
+            case (Bot, _)             => false
```

Comment:
Bot is inf, so its is smaller than anything. So this should be true.

Reply:
Hmm okay I'm confused as to why infinity is smaller than any constant, but I'll accept it :)